pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -210,22 +215,31 @@ func (h *handler) OnRemove(_ string, obj runtime.Object) (runtime.Object, error)
 		return obj, err
 	}
 
-	obj, err := h.run(obj, false)
+	d, err := data.Convert(obj.DeepCopyObject())
 	if err != nil {
 		return nil, err
 	}
 
-	meta, err := meta.Accessor(obj)
+	infraName := d.String("metadata", "name")
+	if !d.Bool("status", "jobComplete") && d.String("status", "failureReason") == "" {
+		return obj, fmt.Errorf("cannot delete machine %s because create job has not finished", infraName)
+	}
+
+	obj, err = h.run(obj, false)
 	if err != nil {
 		return nil, err
 	}
 
-	job, err := h.jobs.Get(meta.GetNamespace(), getJobName(meta.GetName()))
+	job, err := h.jobs.Get(d.String("metadata", "namespace"), getJobName(infraName))
 	if err != nil {
 		return nil, err
 	}
 
-	if condition.Cond("Failed").IsTrue(job) || job.Status.CompletionTime != nil {
+	if job.Status.CompletionTime != nil {
+		// Delete objects owned by infrastructure object.
+		if err := h.apply.WithOwner(obj).ApplyObjects(); err != nil {
```

Comment:
Do we have an explicit call we can change to? This is not intuitive that it will go delete things because it's empty.

Reply:
There doesn't seem to be any other calls. I added a more explanatory comment here.